dead-code-lint: de-dup multiple unused assoc functions

Ezrashaw opened the pull request:
Fixes #109600

Prior art: #97853

Ezrashaw committed:
refactor: emit "unused assoc fn" in lexical order

with repect to other dead code lints

Ezrashaw committed:
dead-code-lint: de-dup multiple unused assoc fns

lcnr commented on compiler/rustc_passes/src/dead.rs:
```diff
@@ -836,6 +836,13 @@ fn check_mod_deathness(tcx: TyCtxt<'_>, module: LocalDefId) {
+            for item in impl_item.items {
+                visitor.check_definition(item.id.owner_id.def_id);
+            }
+            continue;
```
this continue feels unnecessary, but also isn't really harmful :shrug:

lcnr commented on compiler/rustc_passes/src/dead.rs:
```diff
@@ -837,9 +840,26 @@ fn check_mod_deathness(tcx: TyCtxt<'_>, module: LocalDefId) {
-                visitor.check_definition(item.id.owner_id.def_id);
+                match item.kind {
+                    hir::AssocItemKind::Const | hir::AssocItemKind::Type => {
+                        visitor.check_definition(item.id.owner_id.def_id)
```
would it make sense to do this for all associated items, and then change the message to "associated items are not used" if it isn't just methods?

lcnr reviewed (commented):
nits, then r=me

Ezrashaw committed:
impl reviewer feedback

- remove unused (pun intentional) `continue`
- improve wording with assoc items in general

Ezrashaw commented:
@lcnr Implemented.

Note that this lint only applies to inherent `impl`s and therefore, associated types here are kind of useless without the incomplete `inherent_associated_types` feature.

lcnr commented:
@bors r+ rollup

bors commented:
:pushpin: Commit 39e23ef532fa894f33ffcd105e630955bf315411 has been approved by `lcnr`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=lcnr 39e23ef532fa894f33ffcd105e630955bf315411 -->
<!-- homu: {"type":"Approved","sha":"39e23ef532fa894f33ffcd105e630955bf315411","approver":"lcnr","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->

Ezrashaw committed:
bless the single test

Ezrashaw commented:
@lcnr Sorry, I forgot to bless a single test... dammit.

On the other hand, that test's output has objectively improved.

lcnr commented:
@bors r+

bors commented:
:pushpin: Commit 2bafc0fcee7c2ef0ea8c8a8af112e07a83304f79 has been approved by `lcnr`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=lcnr 2bafc0fcee7c2ef0ea8c8a8af112e07a83304f79 -->
<!-- homu: {"type":"Approved","sha":"2bafc0fcee7c2ef0ea8c8a8af112e07a83304f79","approver":"lcnr","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->
